File: pubnub-gson/pubnub-gson-api/src/main/java/com/pubnub/api/java/models/consumer/datasync/entity/PNJsonPatchOperation.java
package com.pubnub.api.java.models.consumer.datasync.entity;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.ToString;
import lombok.experimental.Accessors;
import org.jetbrains.annotations.Nullable;

/**
 * A single JSON Patch (RFC-6902) operation used by {@code pubnub.dataSync().entity().patch()}.
 */
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@Accessors(chain = true)
@ToString
public class PNJsonPatchOperation {
    /**
     * The operation to perform (e.g. {@code add}, {@code remove}, {@code replace}, {@code move}, {@code copy}, {@code test}).
     */
    private String op;

    /**
     * JSON Pointer (RFC-6901) to the target location, e.g. {@code /status} or {@code /payload/role}.
     */
    private String path;

    /**
     * The value to add, replace, or test. Required for {@code add}/{@code replace}/{@code test}.
     */
    @Nullable
    private Object value;

    /**
     * JSON Pointer to the source location. Required for {@code move}/{@code copy}.
     */
    @Nullable
    private String from;
}

File: pubnub-kotlin/pubnub-kotlin-api/src/commonMain/kotlin/com/pubnub/api/models/consumer/datasync/entity/PNJsonPatchOperation.kt
package com.pubnub.api.models.consumer.datasync.entity

/**
 * A single JSON Patch (RFC-6902) operation used by `pubnub.dataSync.updateEntity`.
 *
 * @property op The operation to perform (e.g. `add`, `remove`, `replace`, `move`, `copy`, `test`).
 *   Modeled as a [String] rather than an enum to stay source-compatible with future values.
 * @property path JSON Pointer (RFC-6901) to the target location, e.g. `/status` or `/payload/role`.
 * @property value The value to add, replace, or test. Required for `add`/`replace`/`test`.
 * @property from JSON Pointer to the source location. Required for `move`/`copy`.
 */
data class PNJsonPatchOperation(
    val op: String,
    val path: String,
    val value: Any? = null,
    val from: String? = null,
)
